### PR TITLE
pause until debugger thread resumed so we can debug startup

### PR DESCRIPTION
**Pull request by mykmelez**
@jryans This works as expected when applied on top of #40. Is it something that DevTools might be willing take upstream? Perhaps in coordination with another flag like `--pause-jsdebugger-until-server-thread-resumes-even-if-that-means-hanging-the-app`?

Note: I do see two instances of this error after the server connects:

>Handler function threw an exception: TypeError: this._progressListener is undefined
>Stack: _onDocShellCreated/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/server/actors/webbrowser.js:1319:9
>exports.makeInfallible/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/ThreadSafeDevToolsUtils.js:101:14
>Line: 1319, column: 9

### CODE REVIEW

**Comment by mykmelez** [comment]
Upon further reflection, I might be able to simply delay parsing the *appPath* argument, and starting the app, until `devtools-thread-resumed`. Then I wouldn't need to spin the event loop.

**Comment by mykmelez** [comment]
> Upon further reflection, I might be able to simply delay parsing the appPath argument, and starting the app, until devtools-thread-resumed. Then I wouldn't need to spin the event loop.

I tried this over in #40, and it works, but it has its own drawbacks. I'm unsure which implementation I prefer.

**Comment by jryans** [comment]
This approach to startup debugging seems better to me, since it would be available to any application. I applied this to a Firefox build, and it enabled startup debugging over there as well, which is great!

I think it should be fine to upstream this mostly as-is, but @ochameau and I would need to work out how to handle the `_progressListener` error. This is happening because we currently [expect a window to exist][1] at startup if there are ever going to be windows opening at all. @ochameau and I can work that out in a bug to merge this upstream.

The `_progressListener` is there to notice if windows change location, etc. which is not too important for Positron's platform debugging. So for Positron at least, you could silence the error if you like by changing `webbrowser.js` to test for `this._progressListener` first [where the error happens][2]. That should be enough to merge this approach now for Positron without waiting on upstream.

[2]: https://dxr.mozilla.org/mozilla-central/rev/a884b96685aa13b65601feddb24e5f85ba861561/devtools/server/actors/webbrowser.js#1319

[1]: https://dxr.mozilla.org/mozilla-central/rev/a884b96685aa13b65601feddb24e5f85ba861561/devtools/server/actors/webbrowser.js#1195

**Comment by mykmelez** [comment]
> The _progressListener is there to notice if windows change location, etc. which is not too important for Positron's platform debugging. So for Positron at least, you could silence the error if you like by changing webbrowser.js to test for this._progressListener first where the error happens. That should be enough to merge this approach now for Positron without waiting on upstream.

Ok, sounds good, I'll make that change!

**Comment by mykmelez** [comment]
> Ok, sounds good, I'll make that change!

I made the change in 4735aeb.

File: devtools/server/actors/webbrowser.js [hunk 1]
```diff
@@ -1316,7 +1316,12 @@ TabActor.prototype = {
       // In child processes, we have new root docshells,
       // let's watch them and all their child docshells.
       if (this._isRootDocShell(docShell)) {
-        this._progressListener.watch(docShell);
+        // Don't assume there's a progress listener, since Positron starts
+        // without a window, so this._progressListener never gets created.
+        // TODO: ensure there's a listener for apps without an initial window.
+        if (this._progressListener) {
```
Comment (jryans) [inline devtools/server/actors/webbrowser.js]: I spotted a few more cases where you might want the same thing:

https://dxr.mozilla.org/mozilla-central/rev/f3f2fa1d7eed5a8262f6401ef18ff8117a3ce43e/devtools/server/actors/webbrowser.js#1397
https://dxr.mozilla.org/mozilla-central/rev/f3f2fa1d7eed5a8262f6401ef18ff8117a3ce43e/devtools/server/actors/webbrowser.js#1478

**Comment by jryans** [comment]
Overall this looks good to me! You may want to add the extra checks I just commented for safety, but I wasn't sure whether they'd be hit in Positron or not.

**Comment by mykmelez** [comment]
> Overall this looks good to me! You may want to add the extra checks I just commented for safety, but I wasn't sure whether they'd be hit in Positron or not.

I hit one of them earlier this morning!

> JavaScript error: resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/server/actors/webbrowser.js, line 1402: TypeError: this._progressListener is undefined

(Line number slightly different, presumably because of the changes I've already made to this file, but this is the same as the first additional case you noted.)

Additional cases fixed in d6e3096.

**Comment by jryans** [comment]
Great, this looks ready to go!

**Comment by jryans** [comment]
I'll file a bug to upstream in a bit.

**Comment by mykmelez** [comment]
> I'll file a bug to upstream in a bit.

Thanks, @jryans! Please cc: me on it!

**Comment by jryans** [comment]
This will be merged as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1275942.

**Pull request merged**